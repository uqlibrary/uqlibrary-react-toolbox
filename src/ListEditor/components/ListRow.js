import React, {Component} from 'react';
import PropTypes from 'prop-types';
import FontIcon from 'material-ui/FontIcon';
import IconButton from 'material-ui/IconButton';
import {ConfirmDialogBox} from '../../ConfirmDialogBox';

export default class ListRow extends Component {
    static propTypes = {
        index: PropTypes.number.isRequired,
        item: PropTypes.any.isRequired,
        canMoveUp: PropTypes.bool,
        canMoveDown: PropTypes.bool,
        onMoveUp: PropTypes.func,
        onMoveDown: PropTypes.func,
        onDelete: PropTypes.func,
        locale: PropTypes.object,
        disabled: PropTypes.bool,
        hideReorder: PropTypes.bool
    };

    static defaultProps = {
        locale: {
            moveUpHint: 'Move item up the order',
            moveDownHint: 'Move item down the order',
            deleteHint: 'Remove this item',
            deleteRecordConfirmation: {
                confirmationTitle: 'Delete item',
                confirmationMessage: 'Are you sure you want to delete this item?',
                cancelButtonLabel: 'No',
                confirmButtonLabel: 'Yes'
            }
        }
    };

    constructor(props) {
        super(props);
    }

    showConfirmation = () => {
        this.confirmationBox.showConfirmation();
    }

    deleteRecord = () => {
        if (!this.props.disabled && this.props.onDelete) this.props.onDelete(this.props.item, this.props.index);
    }

    onMoveUp = () => {
        if (!this.props.disabled && this.props.onMoveUp) this.props.onMoveUp(this.props.item, this.props.index);
    }

    onMoveDown = () => {
        if (!this.props.disabled && this.props.onMoveDown) this.props.onMoveDown(this.props.item, this.props.index);
    }

    render() {
        return (
            <div className="columns is-gapless is-mobile listRow datalist datalist-row">
                <ConfirmDialogBox
                    onRef={ref => (this.confirmationBox = ref)}
                    onAction={this.deleteRecord}
                    locale={this.props.locale.deleteRecordConfirmation}/>
                <div className="column datalist-text">
                    {
                        this.props.item.value ? this.props.item.value : this.props.item
                    }
                </div>
                <div className="column is-narrow is-hidden-mobile listReorder datalist-buttons">
                    {!this.props.hideReorder && this.props.canMoveUp &&
                    <IconButton
                        tooltip={this.props.locale.moveUpHint}
                        onTouchTap={this.onMoveUp}
                        className="reorderUp"
                        disabled={this.props.disabled}>
                        <FontIcon className="material-icons">keyboard_arrow_up</FontIcon>
                    </IconButton>
                    }
                    {!this.props.hideReorder && this.props.canMoveDown &&
                    <IconButton
                        tooltip={this.props.locale.moveDownHint}
                        onTouchTap={this.onMoveDown}
                        className="reorderDown"
                        disabled={this.props.disabled}>
                        <FontIcon className="material-icons">keyboard_arrow_down</FontIcon>
                    </IconButton>
                    }
                </div>
                <div className="column is-narrow listDelete datalist-buttons">
                    <IconButton
                        className="itemDelete"
                        tooltip={this.props.locale.deleteHint}
                        onTouchTap={this.showConfirmation}
                        disabled={this.props.disabled}>
                        <FontIcon className="material-icons deleteIcon">delete</FontIcon>
                    </IconButton>
                </div>
            </div>
        );
    }
}

