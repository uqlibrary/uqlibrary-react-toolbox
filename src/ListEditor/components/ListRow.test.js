jest.dontMock('./ListRow');

import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ListRow from './ListRow';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import PropTypes from 'prop-types';
import injectTapEventPlugin from 'react-tap-event-plugin';

function setup({index, item,
    canMoveUp, canMoveDown, onMoveUp, onMoveDown, onDelete, disabled}){

    const props = {
        index: index || 0, // PropTypes.number.isRequired,
        item: item || 'one', // PropTypes.object.isRequired,
        canMoveUp: canMoveUp || false, // PropTypes.bool,
        canMoveDown: canMoveDown || false, // PropTypes.bool,
        onMoveUp, // PropTypes.func,
        onMoveDown, // PropTypes.func,
        onDelete, // PropTypes.func,
        disabled: disabled || false // PropTypes.bool
    };

    return shallow(<ListRow {...props} />);
}


beforeAll(() => {
    injectTapEventPlugin();
});

describe('ListRow renders ', () => {
    it('a row with index and item and delete button', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('a row with index and item set, renders reorder buttons, and delete button', () => {
        const wrapper = setup({canMoveUp: true, canMoveDown: true});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('a row with index and item set calls move up function', () => {
        const testFunction = jest.fn();
        const wrapper = setup({canMoveUp: true, onMoveUp: testFunction});
        const button = wrapper.find('IconButton .reorderUp');
        expect(button.length).toBe(1);

        const buttonDown = wrapper.find('IconButton .reorderDown');
        expect(buttonDown.length).toBe(0);

        wrapper.instance().onMoveUp();
        expect(testFunction).toBeCalled();
    });


    it('a row with index and item set calls move down function', () => {
        const testFunction = jest.fn();
        const wrapper = setup({index: 0, canMoveDown: true, onMoveDown: testFunction});

        const button = wrapper.find('IconButton .reorderDown');
        expect(button.length).toBe(1);

        const buttonUp = wrapper.find('IconButton .reorderUp');
        expect(buttonUp.length).toBe(0);

        wrapper.instance().onMoveDown();
        expect(testFunction).toBeCalled;
    });

    it('a row with index and item set calls delete function', () => {
        const testFunction = jest.fn();
        const wrapper = setup({index: 0, onDelete: testFunction});
        expect(toJson(wrapper)).toMatchSnapshot();
        const button = wrapper.find('IconButton .itemDelete');
        expect(button.length).toBe(1);
        wrapper.instance().deleteRecord();
        expect(testFunction).toBeCalled;
    });
});
