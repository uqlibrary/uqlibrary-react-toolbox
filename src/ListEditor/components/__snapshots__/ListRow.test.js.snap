// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListRow renders  a row with index and item and delete button 1`] = `
<div
  className="columns is-gapless is-mobile listRow datalist datalist-row"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to delete this item?",
        "confirmationTitle": "Delete item",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text"
  >
    one
  </div>
  <div
    className="column is-narrow is-hidden-mobile listReorder datalist-buttons"
  />
  <div
    className="column is-narrow listDelete datalist-buttons"
  >
    <IconButton
      className="itemDelete"
      disableTouchRipple={false}
      disabled={false}
      iconStyle={Object {}}
      onTouchTap={[Function]}
      tooltip="Remove this item"
      tooltipPosition="bottom-center"
      touch={false}
    >
      <FontIcon
        className="material-icons deleteIcon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        delete
      </FontIcon>
    </IconButton>
  </div>
</div>
`;

exports[`ListRow renders  a row with index and item set calls delete function 1`] = `
<div
  className="columns is-gapless is-mobile listRow datalist datalist-row"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to delete this item?",
        "confirmationTitle": "Delete item",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text"
  >
    one
  </div>
  <div
    className="column is-narrow is-hidden-mobile listReorder datalist-buttons"
  />
  <div
    className="column is-narrow listDelete datalist-buttons"
  >
    <IconButton
      className="itemDelete"
      disableTouchRipple={false}
      disabled={false}
      iconStyle={Object {}}
      onTouchTap={[Function]}
      tooltip="Remove this item"
      tooltipPosition="bottom-center"
      touch={false}
    >
      <FontIcon
        className="material-icons deleteIcon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        delete
      </FontIcon>
    </IconButton>
  </div>
</div>
`;

exports[`ListRow renders  a row with index and item set, renders reorder buttons, and delete button 1`] = `
<div
  className="columns is-gapless is-mobile listRow datalist datalist-row"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to delete this item?",
        "confirmationTitle": "Delete item",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text"
  >
    one
  </div>
  <div
    className="column is-narrow is-hidden-mobile listReorder datalist-buttons"
  >
    <IconButton
      className="reorderUp"
      disableTouchRipple={false}
      disabled={false}
      iconStyle={Object {}}
      onTouchTap={[Function]}
      tooltip="Move item up the order"
      tooltipPosition="bottom-center"
      touch={false}
    >
      <FontIcon
        className="material-icons"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        keyboard_arrow_up
      </FontIcon>
    </IconButton>
    <IconButton
      className="reorderDown"
      disableTouchRipple={false}
      disabled={false}
      iconStyle={Object {}}
      onTouchTap={[Function]}
      tooltip="Move item down the order"
      tooltipPosition="bottom-center"
      touch={false}
    >
      <FontIcon
        className="material-icons"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        keyboard_arrow_down
      </FontIcon>
    </IconButton>
  </div>
  <div
    className="column is-narrow listDelete datalist-buttons"
  >
    <IconButton
      className="itemDelete"
      disableTouchRipple={false}
      disabled={false}
      iconStyle={Object {}}
      onTouchTap={[Function]}
      tooltip="Remove this item"
      tooltipPosition="bottom-center"
      touch={false}
    >
      <FontIcon
        className="material-icons deleteIcon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        delete
      </FontIcon>
    </IconButton>
  </div>
</div>
`;
